rfcs/0016-const-fn.md

```diff
@@ -0,0 +1,406 @@
+
+The returned log data contains a `Vec<u8>` obtained by using the ABIEncode.
+
+Using VM evaluation we can store the generated ABIEncoded byte array into the
+final bytecode of the program.
+
+To compile the example above it takes 0.76 seconds including core, without core
+it takes less than 0.02 seconds to compile. To bootstrap the VM and run the
+script it takes less than a millisecond.
+Thus the overhead for doing `const fn` evaluation with the VM should be around
+**tens of milliseconds**, which is negligible compared to compiling a program
+with std which takes around 2.5 seconds.
+
+## Enforcing constness rules
+
+To ensure that `const fn`s maintain predictable compile-time behavior, we enforce a set of constness rules that restrict what can and cannot be done inside a `const fn`. These rules prevent operations that rely on runtime state while allowing deterministic computation at compile time.
+
+- No access to storage
+- Cannot use certain asm opcodes
+- Error out when `const context` is assigned to a heap type.
+
+### Non usable asm opcodes
+
+The same opcodes that are disallowed to predicates are also disallowed in const fn:
+
+- BAL
+- BHEI
+- BHSH
+- BURN
+- CALL
+- CB
+- CCP
+- CROO
+- CSIZ
+- GM
+- LOG
+- LOGD
+- MINT
+- RETD
+- SMO
+- SRW
+- SRWQ
+- SWW
+- SWWQ
+- TIME
+- TR
+- TRO
+
+### Heap types
+
+Heap types are a challenge for `const fn` for multiple reasons.
+First is the how to map the binary data section entry for a Heap type to the heap.
+Another challenge is that to properly use heap types such as `Vec<T>` we want to
+support methods such as: `fn push(ref mut self, value: T)`, and this would require
+ `const fn`s to support `ref mut` parameters.
+
+As heap types created in the binary data section would require additional effort to be 
+used at runtime and in small cases this would offset the time required to recompute 
+them. It would be better to disallow `const fn` from returning or receiving Heap 
+types but still allow them to use Heap types inside the `const fn` code blocks.
+
+Thus we propose to not try to serialize and deserialize the heap types in the 
+binary data section instead throw warnings or errors when the compiler tries to store
+the result of a const evaluation in the binary data section.
+
+This allows `const fn`s to use heap types internally but restricts types in the data
+sections to non heap types.
+
+### `ref mut` parameters
+
+Support of `ref mut` in `const fn` is a requirement for properly
+handling heap type such as `Vec<T>` and function such as `fn push(ref mut self, value: T)`.
+
+A `const fn` with a `ref mut` parameters is callable inside other regular and `const fn`s but
+cannot be used directly in other `const contexts` such as constant items.
+
+## Final binary
+
+When a `const fn`is not called at runtime the final binary outputted by the compiler
+won't have any opcodes related to the computation of `const fn`s.
+
```

Comment:
Presumably this is handled by the dead code elimination phase? The UX for this scenario needs to go hand-in-hand with any warnings or whatever when a `const fn` is never called in a const context.